forge/test/operators/pytorch/matmul/test_matmul.py

```diff
@@ -0,0 +1,295 @@
+from loguru import logger
+
+import forge
+import forge.op
+
+from forge.op_repo import TensorShape
+
+from test.operators.utils import InputSourceFlags, VerifyUtils
+from test.operators.utils import FailingReasons
+from test.operators.utils.utils import TestDevice
+from test.operators.utils import ShapeUtils
+from test.operators.utils import PytestParamsUtils
+from test.operators.utils import ValueRanges
+
+
+class ModelFromAnotherOp(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.testname = "Matmul_operator_test_op_src_from_another_op"
+
+    def forward(self, x, y):
+        xx = torch.add(x, x)
+        yy = torch.add(y, y)
+        return torch.matmul(xx, yy)
+
+
+class ModelFromHost(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.testname = "Matmul_operator_test_op_src_from_host"
+
+    def forward(self, x, y):
+        return torch.matmul(x, y)
+
+
+class ModelFromDramQueue(ModelFromHost):
+    def __init__(self):
+        super().__init__()
+        self.testname = "Matmul_operator_test_op_src_from_dram_queue"
+
+
+class ModelConstEvalPass(torch.nn.Module):
+    def __init__(self, shape: TensorShape):
+        super().__init__()
+        self.testname = "Matmul_operator_test_op_src_const_eval_pass"
+        self.c1 = (torch.rand(ShapeUtils.reduce_microbatch_size(shape), requires_grad=False) - 0.5).detach()
+        self.c2 = (
+            torch.rand(ShapeUtils.reduce_microbatch_size(ShapeUtils.switch_last_two(shape)), requires_grad=False) - 0.5
+        ).detach()
+
+    def forward(self, x, y):
+        mm1 = torch.matmul(self.c1, self.c2)
+        mm2 = torch.matmul(x, y)
+        aa = torch.add(mm1, mm2)
+        return aa
+
+
+def verify(
+    test_device: TestDevice,
+    model: Type[torch.nn.Module],
+    input_shape: TensorShape,
+    input_params: List[Dict] = [],
+    input_source_flag: InputSourceFlags = None,
+    dev_data_format: forge.DataFormat = None,
+    math_fidelity: forge.MathFidelity = None,
+):
+    """Common verification function for all tests"""
+
+    pytorch_model = model(input_shape) if model == ModelConstEvalPass else model()
+
+    # generate shapes:
+    input_shapes = [input_shape, ShapeUtils.switch_last_two(input_shape)]
+
+    logger.trace(f"***input_shapes: {input_shapes}")
+
+    VerifyUtils.verify(
+        model=pytorch_model,
+        test_device=test_device,
+        input_shapes=input_shapes,
+        input_params=input_params,
+        input_source_flag=input_source_flag,
+        dev_data_format=dev_data_format,
+        math_fidelity=math_fidelity,
+        value_range=ValueRanges.SMALL,
+    )
+
+
+# Prepare the tests data which is list of pytest.param() objects.
+# First, create the PytestParamsUtils object that holds the test data.
+utils = PytestParamsUtils()
+
+# fmt: off
+
+# Then use this method to generate the test data.
+# It should be called just once to generate all test cases.
+utils.generate_test_params_list(
```

Comment:
Can we add more detailed comments for this like:

"Generate parameters for testing 3D and 4D input shapes for matmul operations
with different model configurations and data format fidelity levels"

Reply:
I add this comment, tnx!